consensus/hotstuff/committees/consensus_committee.go

```diff
@@ -236,6 +296,92 @@ func (c *Consensus) DKG(view uint64) (hotstuff.DKG, error) {
 	return epochInfo.dkg, nil
 }
 
+// handleProtocolEvents is a worker routine which processes protocol events from
+// the protocol state. When we observe a new epoch being committed, we compute
+// the leader selection and cache static info for the epoch. When we observe
+// epoch emergency fallback being triggered, we inject a fallback epoch.
+func (c *Consensus) handleProtocolEvents(ctx irrecoverable.SignalerContext, ready component.ReadyFunc) {
+	ready()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case epoch := <-c.committedEpochsCh:
+			_, err := c.prepareEpoch(epoch)
+			if err != nil {
+				ctx.Throw(err)
+			}
+		case <-c.epochEmergencyFallback:
+			err := c.onEpochEmergencyFallbackTriggered()
+			if err != nil {
+				ctx.Throw(err)
+			}
+		}
+	}
+}
+
+// EpochCommittedPhaseStarted passes the protocol event to the worker thread.
+func (c *Consensus) EpochCommittedPhaseStarted(_ uint64, first *flow.Header) {
+	nextEpoch := c.state.AtBlockID(first.ID()).Epochs().Next()
+	c.committedEpochsCh <- nextEpoch
+}
+
+// EpochEmergencyFallbackTriggered passes the protocol event to the worker thread.
+func (c *Consensus) EpochEmergencyFallbackTriggered() {
+	c.epochEmergencyFallback <- struct{}{}
+}
+
+// onEpochEmergencyFallbackTriggered handles the protocol event for emergency epoch
+// fallback mode being triggered. When this occurs, we inject a fallback epoch
+// to the committee which extends the current epoch.
+// This method must also be called on initialization, if emergency fallback mode
+// was triggered in the past.
+// No errors are expected during normal operation.
+func (c *Consensus) onEpochEmergencyFallbackTriggered() (retErr error) {
```

Comment:
Agreed, it's not a problem in this specific case. However it could also be a problem when we make EECC-triggering fork dependent -> I'll keep that in mind for https://github.com/dapperlabs/flow-go/issues/6316.